### DISCUSSION TITLE
isSubmitSuccessful triggering true on promises before form is submitted

### DISCUSSION THREAD

**Discussion by mahe993** [react-hook-form/react-hook-form#9284] [Q&A]
const { ...., formState } = useForm();
const { isSubmitSuccessful } = formState;
const onSubmit = async (data) => {
   console.log(isSubmitSuccessful); // false
   await new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, 100);
    console.log(isSubmitSuccessful); // true
   Promise.reject("Error");
   console.log(isSubmitSuccessful); // false
  };
  
this is causing unintended re-renders when using useEffect with isSubmitSuccessful as dependency.

**Comment by leapful** [comment]
@mahe993 Please provide a codesandbox url to re-produce the issue.

**Reply by mahe993** [reply]
@leapful  https://codesandbox.io/s/practical-swirles-evok04?file=/src/App.js

**Reply by leapful** [reply]
@mahe993

You implementation for `onValid` callback of method `handleSubmit` is wrong. Your "onSubmit" callback is not an `async await` function therefore when it's passed as `onValid` callback to `handleSubmit` therefore no await on it, leading to wrong behavior of flag `isSubmitSuccessful`.

Please make sure `onValid` callback is an `async await` function and all of its related logics are `await` correctly to achieve the correct behavior of flag `isSubmitSuccessful`.

Working demo: https://codesandbox.io/s/amazing-maxwell-8edrbc?file=/src/App.js

**Reply by mahe993** [reply]
Thank you for this!
